### PR TITLE
[Breaking] Add training to batchnorm with exponential moving average

### PR DESCRIPTION
**Pull request by nkoppel**
Ports the [dfdx implementation](https://github.com/coreylowman/dfdx/blob/main/dfdx/src/nn/layers/batch_norm1d.rs) of batch norm training to candle, which unfortunately requires breaking changes. 

- Changes the type of `running_mean` and `running_var` from `Tensor` to `Var`. I'm not sure if this is an intended use of `Var`, but it seemed to be the best fit for this use case.
- Removes `num_features` parameter and adds `momentum` parameter to `new` and `new_no_bias` methods. The `num_features` parameter is redundant with the size of the given tensors.
- Adds a `forward_t` method that shadows the `ModuleT` implementation. This makes batchnorm more intuitive in a `ModuleT` context, but could create confusion or have unexpected behavior for some use cases. Without this, however, `forward_t(..., true)` calls `forward` instead of `forward_learning`, so I'm not really sure of the trade-off here.

### CODE REVIEW

File: candle-nn/src/batch_norm.rs [hunk 1]
```diff
@@ -142,26 +184,51 @@ impl BatchNorm {
         let x = x.flatten_from(1)?.contiguous()?;
         let x = if self.remove_mean {
             let mean_x = x.mean_keepdim(1)?;
+            {
+                // Update running mean
+                let new_mean = ((self.running_mean.as_tensor() * (1.0 - self.momentum))?
+                    + (mean_x.reshape(((),))? * self.momentum)?)?;
```
Comment (LaurentMazare) [inline candle-nn/src/batch_norm.rs:190]: `flatten_all`?

File: candle-nn/src/batch_norm.rs [hunk 2]
```diff
@@ -34,79 +42,108 @@ impl From<f64> for BatchNormConfig {
...
         weight: Tensor,
         bias: Tensor,
         eps: f64,
+        momentum: f64,
     ) -> Result<Self> {
         if eps < 0. {
             candle::bail!("batch-norm eps cannot be negative {eps}")
         }
-        if weight.dims() != [num_features] {
+        if running_mean.dims().len() != 1 {
```
Comment (LaurentMazare) [inline candle-nn/src/batch_norm.rs:72]: `rank`?

File: candle-nn/src/batch_norm.rs [hunk 3]
```diff
@@ -34,79 +42,108 @@ impl From<f64> for BatchNormConfig {
...
 
 impl BatchNorm {
     pub fn new(
-        num_features: usize,
         running_mean: Tensor,
         running_var: Tensor,
         weight: Tensor,
         bias: Tensor,
         eps: f64,
+        momentum: f64,
```
Comment (LaurentMazare) [inline candle-nn/src/batch_norm.rs:67]: Let's have a separate function to create a batch-norm with momentum and use the default value here so as not to break the current api (also most use cases don't care much about the momentum value anyway).

File: candle-nn/src/batch_norm.rs [hunk 4]
```diff
@@ -142,26 +184,51 @@ impl BatchNorm {
...
+            }
             x.broadcast_sub(&mean_x)?
         } else {
             x
         };
         let norm_x = x.sqr()?.mean_keepdim(1)?;
+        {
+            // Update running variance
+            let running_var_weight = 1.0 - self.momentum;
+            let norm_x_weight = self.momentum * num_features as f64 / (num_features as f64 - 1.0);
```
Comment (LaurentMazare) [inline candle-nn/src/batch_norm.rs:202]: I'm a bit confused by this, shouldn't this be the batch-size rather than the number of features to get the unbiased estimator?
But that would also mean that the two `mean_keepdim` above should be done on the first dimension (batch dim) rather than on the second (channels)?
What testing did you run on this?

File: candle-nn/src/batch_norm.rs [hunk 5]
```diff
@@ -34,79 +42,108 @@ impl From<f64> for BatchNormConfig {
             eps,
             remove_mean: true,
             affine: true,
+            momentum: 0.1,
         }
     }
 }
 
-#[derive(Clone, Debug)]
+#[derive(Debug)]
```
[the remark this answers is not in the document]
Comment (nkoppel) [reply candle-nn/src/batch_norm.rs]: This was removed due to an earlier effort to implement this with `Rwlock<Tensor>` instead of `Var`, since I don't know candle all that well yet.

File: candle-nn/src/batch_norm.rs (hunk 3 above)
Comment (nkoppel) [reply candle-nn/src/batch_norm.rs]: Okay, I've added `with_momentum` variants of the `new` methods. What are your thoughts on my removing of the `num_features` parameter from the `new` methods? It does simplify things, but is still breaking. These changes will break some code due to the changes to `BatchNormConfig`.

File: candle-nn/src/batch_norm.rs (hunk 4 above)
Comment (nkoppel) [reply candle-nn/src/batch_norm.rs]: This should be resolved now, and I have added to the batchnorm test to verify the new behavior against pytorch.

File: candle-nn/src/batch_norm.rs (hunk 3 above)
Comment (LaurentMazare) [reply candle-nn/src/batch_norm.rs]: Would prefer not breaking the api and so adding `num_features` back.

Comment (nkoppel) [reply candle-nn/src/batch_norm.rs]: I've added it back, and modified `check_validity` to use it.

**Review by LaurentMazare — APPROVED** [review]

**Comment by LaurentMazare** [comment]
Thanks!
